keras_cv/models/mlp_mixer.py

```diff
@@ -0,0 +1,279 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""MLP Mixer models for KerasCV.
+
+Reference:
+  - [MLP-Mixer: An all-MLP Architecture for Vision](https://arxiv.org/abs/2105.01601)
+"""
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import backend
+from tensorflow.keras import layers
+
+
+def MLPBlock(mlp_dim, name=None):
+    """An MLP block consisting of two linear layers with GELU activation in
+    between.
+
+    Args:
+      mlp_dim: integer, the number of units to be present in the first layer.
+      name: string, block label.
+
+    Returns:
+      a function that takes an input Tensor representing an MLP block.
+    """
+    if name is None:
+        name = f"mlp_block_{backend.get_uid('mlp_block')}"
+
+    def apply(x):
+        y = layers.Dense(mlp_dim, name=f"{name}_dense_1")(x)
+        y = layers.Activation("gelu", name=f"{name}_gelu")(y)
+        return layers.Dense(x.shape[-1], name=f"{name}_dense_2")(y)
+
+    return apply
+
+
+def MixerBlock(tokens_mlp_dim, channels_mlp_dim, name=None):
+    """A mixer block.
+
+    Args:
+      tokens_mlp_dim: integer, number of units to be present in the MLP block
+        dealing with tokens.
+      channels_mlp_dim: integer, number of units to be present in the MLP block
+        dealing with channels.
+      name: string, block label.
+
+    Returns:
+      a function that takes an input Tensor representing an MLP block.
+    """
+    if name is None:
+        name = f"mixer_block_{backend.get_uid('mlp_block')}"
+
+    def apply(x):
+        y = layers.LayerNormalization()(x)
+        y = layers.Permute((2, 1))(y)
+
+        y = MLPBlock(tokens_mlp_dim, name=f"{name}_token_mixing")(y)
+        y = layers.Permute((2, 1))(y)
+        x = layers.Add()([x, y])
+
+        y = layers.LayerNormalization()(x)
+        y = MLPBlock(channels_mlp_dim, name=f"{name}_channel_mixing")(y)
+        return layers.Add()([x, y])
+
+    return apply
+
+
+def MLPMixer(
+    patch_size,
+    num_blocks,
+    hidden_dim,
+    tokens_mlp_dim,
+    channels_mlp_dim,
+    include_rescaling,
+    include_top,
+    classes=None,
+    weights=None,
+    input_shape=(224, 224, 3),
```

Comment:
Hey Sayak, Is there a reason for (224, 224, 3) other than it being the imagenet size?

Reply:
Yes, it's a sensible default that's why. Could have been (384, 384) too. But we need non `None` spatial resolutions here as there's a `Reshape()` operation dependent on it.

Reply:
Gotcha, I wonder if it makes sense to have this be a required argument instead of providing a default.

Reply:
I suppose it depends on how we anticipate the architecture to be used!

Reply:
I provided a sensible default. Having it as a required argument won't hurt anything I guess.

Reply:
> I suppose it depends on how we anticipate the architecture to be used!

I won't expect this architecture to do good at tasks other than image classification. All the other tasks such as segmentation, detection benefit from hierarchical features but MLP-Mixer is an isotropic architecture meaning it maintains a single resolution of the feature maps throughout its existence.

Reply:
lets make sure to discuss how to compute some alternative values for valid configurations in the docstrings.